doc/source/development/rfc/rfc104_gdal_cli.rst

```diff
@@ -0,0 +1,1175 @@
+    Positional arguments:
+
+    Common Options:
+      -h, --help    Display help message and exit
+      --json-usage  Display usage as JSON document and exit
+      --progress    Display progress bar
+
+    <PIPELINE> is of the form: read [READ-OPTIONS] ( ! <STEP-NAME> [STEP-OPTIONS] )* ! write [WRITE-OPTIONS]
+
+    Example: 'gdal vector pipeline --progress ! read in.gpkg ! \
+                   reproject --dst-crs=EPSG:32632 ! write out.gpkg --overwrite'
+
+    Potential steps are:
+
+    * read [OPTIONS] <INPUT>
+    ------------------------
+
+    Read a vector dataset.
+
+    Positional arguments:
+      -i, --input <INPUT>                                  Input vector dataset [required]
+
+    Options:
+      -l, --layer, --input-layer <INPUT-LAYER>             Input layer name(s) [may be repeated]
+
+    Advanced Options:
+      --if, --input-format <INPUT-FORMAT>                  Input formats [may be repeated]
+      --oo, --open-option <KEY=VALUE>                      Open options [may be repeated]
+
+    * filter [OPTIONS]
+    ------------------
+
+    Filter.
+
+    Options:
+      --bbox <BBOX>                                        Bounding box as xmin,ymin,xmax,ymax [may be repeated]
+
+    * reproject [OPTIONS]
+    ---------------------
+
+    Reproject.
+
+    Options:
+      -s, --src-crs <SRC-CRS>                              Source CRS
+      -d, --dst-crs <DST-CRS>                              Destination CRS [required]
+
+    * write [OPTIONS] <OUTPUT>
+    --------------------------
+
+    Write a vector dataset.
+
+    Positional arguments:
+      -o, --output <OUTPUT>                                Output vector dataset [required]
+
+    Options:
+      -f, --of, --format, --output-format <OUTPUT-FORMAT>  Output format
+      --co, --creation-option <KEY=VALUE>                  Creation option [may be repeated]
+      --lco, --layer-creation-option <KEY=VALUE>           Layer creation option [may be repeated]
+      --overwrite                                          Whether overwriting existing output is allowed
+      --update                                             Whether updating existing dataset is allowed
+      --overwrite-layer                                    Whether overwriting existing layer is allowed
+      --append                                             Whether appending to existing layer is allowed
+      -l, --output-layer <OUTPUT-LAYER>                    Output layer name
+
+
+CLI specification
+-----------------
+
+Subcommand syntax
++++++++++++++++++
+
+.. code-block:: shell
+
+        gdal <subcommand> [<subsubcommand>]... [<options>]... [<positional arguments>]...
+
+where subcommand is something like ``raster``, ``vector``, etc. with potential
+sub-subcommand like ``info``, ``convert``, etc.
+
+Option naming conventions
++++++++++++++++++++++++++
+
+* One-letter short names preceded with dash: ``-z``
+
+  When a value is specified, it must be separated with a space: ``-z <value>``
```

Comment:
Why is "one-letter" only accepting the space, and the longer names also accepts equal sign? I would expect the same criteria for both.

Reply:
I believe this is kind of a "standard" for Unix CLI. Typically if a one-letter option is a boolean, there are Unix utilities (like tar), where doing "-xyz" is interpreted as "-x -y -z", but I haven't developed that mode

Trying with curl:
``curl -H "foo:bar" https://example.com`` works as expected.
But ``curl -H="foo:bar" https://example.com`` sends ``=foo:bar"`` as a header
That said the long version requires a space separator too: ``curl --header "foo:bar" https://example.com``

Reply:
Oh, that is like `make -j8` and `make -j 8`. It allows to "join" the `j` and the number (also confusing to me).

It is probably from the time where 640 kB should be enough for everybody :P

Reply:
oh no, `getopt` is way older than that. it's from a time when the maximum memory was 64 kByte. GNU getopt is the modern version of that.